# Filippo Gastaldello - 03/07/2027

# Generate mutated protein coding sequences for each haplotype in the haplotype csv file

library(tidyverse)
library(parallel)
library(Biostrings)

# FUNCTIONS

# Functions are placed first to let snakemake know they exist before running the code.

build_mutated_sequence <- function(wt_sequence, transcript_variants, haplotype) {
        
        sequence <- character()
        variants <- get_variant_df(data.frame('variant_types' = strsplit(transcript_variants[1,"variant_types"], ","),
                                              'nucleotide_change' = strsplit(transcript_variants[1,"nucleotide_change"], ",")
                                              )
                                   )
        # Keep track of start and stop codons alteration
        start_gain <- FALSE; multiple_starts_in_5 <- FALSE; stop_loss_3_translated <- FALSE
        # Exclude intronic variants ('+' in nucleotide change)
        variants <- variants %>% dplyr::filter(!str_detect(nucleotide_change, "\\+"))
        # Filter variants in the 5' UTR
        variants_5utr <- variants %>% dplyr::filter(str_detect(nucleotide_change, "-"))
        # Remove variants that are not within the 5'utr
        variants_5utr <- variants_5utr %>% dplyr::filter(pos<nchar(wt_sequence$`5utr`))
        # 5' UTR is only needed if a start codon is gained, so apply modification and check for 'ATG' in the resulting sequence
        if (nrow(variants_5utr)>0) {
                # Get 5' UTR sequence
                utr5 <- wt_sequence$`5utr`
                
                SNPs <- variants_5utr %>% filter(desc=="snp")
                DELs <- variants_5utr %>% filter(desc=="del")
                INs <- variants_5utr %>% filter(desc=="ins")
                
                # Compute shift vector 
                shift <- rep(0, nchar(utr5))
                if (nrow(DELs)+nrow(INs) > 0) {
                        shift <- compute_shift(shift, INs, DELs)
                }
                
                # Apply SNPs first
                if (nrow(SNPs)>0) {
                        utr5 <- apply_snp(utr5, SNPs, TRUE)
                }
                # Apply deletions
                if (nrow(DELs)>0) {
                        utr5 <- apply_deletion(utr5, DELs, TRUE)
                }
                # Apply insertions
                if (nrow(INs)>0) {
                        utr5 <- apply_insertion(utr5, INs, TRUE)
                }
                # Remove deletions placeholders
                utr5 <- gsub("D","",utr5)
                # Check if there are start codons in the 5' UTR
                if (str_detect(utr5, "ATG")) {
                        # We need to distinguish 'ATG' that were already there from those generated by the mutations
                        
                        # Get all the starting codons already present in the wt sequence
                        wt_starts <- as_data_frame(str_locate_all(wt_sequence$`5utr`, "ATG")[[1]])
                        # Compute start position wrt shift
                        wt_starts <- wt_starts %>% mutate(wt_shifted_start = start + shift[start]) %>% dplyr::select(wt_shifted_start)
                        # Find the starts in the mutated sequence
                        starts <- as_data_frame(str_locate_all(utr5, "ATG")[[1]]) %>% dplyr::select(start)
                        # Flag multiple starts
                        multiple_starts_in_5 <- TRUE
                        #only keep those that are new
                        starts <- starts %>% left_join(wt_starts, join_by(start==wt_shifted_start), keep=TRUE)
                        starts <- starts %>% mutate(wt_shifted_start = ifelse(is.na(wt_shifted_start),0,wt_shifted_start)) %>% filter(!start==wt_shifted_start)
                        # if something remains it is a gained start codon and the 5' should be used in the translation
                        if (nrow(starts)>0) {
                                starts <- starts %>% arrange(start)
                                # cut from the first start codon and add to sequence
                                utr5 <- str_sub(utr5, as.numeric(starts[1,"start"]), nchar(utr5))
                                # Add 5utr to the sequence
                                sequence <- utr5
                                # flag start gain
                                start_gain <- TRUE
                        } # ELSE nothing to do, 5' UTR do not need to be added to the sequence, move on with cds generation
                }
        }
        # CDS need to be modified only if there are cds non synonymous variants
        cds_variants <- variants %>% filter(!str_detect(nucleotide_change, "\\*"),
                                            !str_detect(nucleotide_change, "-"))
        if(nrow(cds_variants)>0){
                # Get cds sequence
                cds <- wt_sequence$coding
                
                SNPs <- cds_variants %>% filter(desc=="snp")
                DELs <- cds_variants %>% filter(desc=="del")
                INs <- cds_variants %>% filter(desc=="ins")
                
                # Apply SNPs first
                if (nrow(SNPs)>0) {
                        cds <- apply_snp(cds, SNPs, FALSE)
                }
                # Apply deletions
                if (nrow(DELs)>0) {
                        cds <- apply_deletion(cds, DELs, FALSE)
                }
                # Apply insertions
                if (nrow(INs)>0) {
                        cds <- apply_insertion(cds, INs, FALSE)
                }
                # Remove deletions placeholders
                cds <- gsub("D","",cds)
                # If the 5' UTR was added to the sequence, just add the cds part
                if (is_empty(sequence)) {
                        sequence <- paste0(sequence,cds)
                }else{
                        # If 5'UTR was not used, find the fisrt start codon, cut the cds from there and add it to the sequence
                        start <- as_data_frame(str_locate(cds, "ATG"))
                        if (nrow(start)>0) {
                                sequence <- str_sub(cds,start[1,"start"], nchar(cds))
                        }# If no start codon is found and 5' was not use, don't add anything to the sequence
                        
                }
        }else{  # If there are no variants in the CDS include it in the sequence without modifications
                sequence <- paste0(sequence, wt_sequence$coding)
        }
        # 3' UTR is only needed if there are no stop codons in the sequence generated up to this point
        if (!str_detect(as.character(translate(DNAString(sequence))), "\\*")) {
                # Need to add 3'UTR
                
                # Flag 3' translation
                stop_loss_3_translated <- TRUE
                # Check if there are variants in the 3'UTR
                variants_3utr <- variants %>% filter(str_detect(nucleotide_change, "\\*"))
                if (nrow(variants_3utr)>0) {
                        # Modify and add
                        # Get cds sequence
                        utr3 <- wt_sequence$`3utr`
                        
                        SNPs <- variants_3utr %>% filter(desc=="snp")
                        DELs <- variants_3utr %>% filter(desc=="del")
                        INs <- variants_3utr %>% filter(desc=="ins")
                        
                        # Apply SNPs first
                        if (nrow(SNPs)>0) {
                                utr3 <- apply_snp(utr3, SNPs, FALSE)
                        }
                        # Apply deletions
                        if (nrow(DELs)>0) {
                                utr3 <- apply_deletion(utr3, DELs, FALSE)
                        }
                        # Apply insertions
                        if (nrow(INs)>0) {
                                utr3 <- apply_insertion(utr3, INs, FALSE)
                        }
                        # Remove deletion placeholders
                        utr3 <- gsub("D", "", utr3)
                        # Add UTR to sequence
                        sequence <- paste0(sequence, utr3)
                }else{
                        # Add without modifications
                        sequence <- paste0(sequence, wt_sequence$`3utr`)
                }
        }
        
        return(data.frame("nn_sequence" = sequence,
                          "start_gain" = start_gain,
                          "multiple_starts_in_5" = multiple_starts_in_5,
                          "stop_loss_3_translated" = stop_loss_3_translated)
               )
}

get_variant_df <- function(variants){
        
        colnames(variants) <- c("variant_types", "nucleotide_change")
        
        # Process variants one at a time
        res <- lapply(seq(1, nrow(variants)), function(row){
                variant <- variants[row,]
                
                # Treat each variant type independently
                if (str_detect(variant$nucleotide_change, "del")) {
                        # DELETION
                        # Need to understand which separator to use (_* for 3', _- for 5', _ for cds)
                        separator <- get_indel_separator(variant$nucleotide_change)
                        # Get variant reference and alternative alleles
                        ref <- str_split_i(variant$nucleotide_change, "del", 2)
                        alt <- ""
                        # Get variant position
                        pos <- str_split_i(str_split_i(variant$nucleotide_change, separator, 2), "del", 1)
                        if (is.na(pos)) { # (handling of 1 base deletions)
                                separator <- get_separator(variant$nucleotide_change)
                                pos <- str_split_i(str_split_i(variant$nucleotide_change, separator, 2), "del", 1)
                        }else{
                                if(str_detect(variant$nucleotide_change, "-")){
                                        pos <- as.numeric(pos) + as.numeric(nchar(ref)) -1
                                }else{
                                        pos <- as.numeric(pos) - as.numeric(nchar(ref)) +1
                                }
                        }
                        return(data.frame("variant_type"=variant$variant_types,
                                          "nucleotide_change"=variant$nucleotide_change,
                                          "pos"=as.numeric(pos),
                                          "ref"=ref,
                                          "alt"=alt,
                                          "desc"="del"))
                }else if (str_detect(variant$nucleotide_change, "ins")) {
                        # INSERTION
                        separator <- get_indel_separator(variant$nucleotide_change)
                        pos <- str_split_i(str_split_i(variant$nucleotide_change, separator, 2), "ins", 1)
                        # Handle negative index in 5'
                        if(str_detect(variant$nucleotide_change,"-")){
                                pos <- as.numeric(pos)+1
                        }
                        ref <- ""
                        alt <- str_split_i(variant$nucleotide_change, "ins", 2)
                        return(data.frame("variant_type"=variant$variant_types,
                                          "nucleotide_change"=variant$nucleotide_change,
                                          "pos"=as.numeric(pos),
                                          "ref"=ref,
                                          "alt"=alt,
                                          "desc"="ins"))
                }else if (str_detect(variant$nucleotide_change, "dup")) {
                        # DUPLICATION (treat duplications as insertions)
                        ref <- ""
                        alt <- str_split_i(variant$nucleotide_change, "dup", 2)
                        if(nchar(alt)>1){
                                separator <- get_indel_separator(variant$nucleotide_change)
                        }else{
                                separator <- get_separator(variant$nucleotide_change)
                        }
                        pos <- str_split_i(str_split_i(variant$nucleotide_change, separator, 2), "dup", 1)
                        if(str_detect(variant$nucleotide_change, "-")){
                                pos <- as.numeric(pos) + as.numeric(nchar(alt)) -1
                        }else{
                                pos <- as.numeric(pos) - as.numeric(nchar(alt)) +1
                        }
                        return(data.frame("variant_type"=variant$variant_types,
                                          "nucleotide_change"=variant$nucleotide_change,
                                          "pos"=as.numeric(pos),
                                          "ref"=ref,
                                          "alt"=alt,
                                          "desc"="ins"))
                }else{
                        # SNP
                        # Need to understand which separator to use (* for 3', - for 5', . for cds)
                        separator <- get_separator(variant$nucleotide_change)
                        pos <- str_sub(str_split_i(str_split_i(variant$nucleotide_change, separator, 2), ">",1),
                                       1,
                                       nchar(str_split_i(str_split_i(variant$nucleotide_change, separator, 2), ">",1))-1)
                        ref <- str_split_i(str_split_i(variant$nucleotide_change, pos, 2),">",1)
                        alt <- str_split_i(variant$nucleotide_change, ">", 2)
                        return(data.frame("variant_type"=variant$variant_types,
                                          "nucleotide_change"=variant$nucleotide_change,
                                          "pos"=as.numeric(pos),
                                          "ref"=ref,
                                          "alt"=alt,
                                          "desc"="snp"))
                }
        })

        return(bind_rows(res))
        
}

compute_shift <- function(shift, INs, DELs){
        # MANAGE INSERTIONS
        if(nrow(INs)>0){
                for (IN in seq(1, nrow(INs))) {
                        
                        start <- length(shift) - as.numeric(INs[IN,"pos"])
                        shift <- c(shift[1: start], shift[(start+1): (length(shift))]+rep(nchar(INs[IN, "alt"]), length(shift)-start))
                }
        }
        # MANAGE DELETIONS
        if(nrow(DELs)>0){
                for (DEL in seq(1, nrow(DELs))) {
                        start <- length(shift) - as.numeric(DELs[DEL,"pos"])
                        end <-  start + nchar(DELs[DEL,"ref"]) + 1
                        # Check if deletion happens at the end of the sequence
                        if (end > length(shift)) {
                                shift <- c(shift[1:start],
                                           rep(NA, as.numeric(nchar(DELs[DEL,"ref"]))))        
                        }else{
                                shift <- c(shift[1:start],
                                           rep(NA, as.numeric(nchar(DELs[DEL,"ref"]))),
                                           shift[end: length(shift)]-rep(nchar(DELs[DEL,"ref"]), length(shift)-end+1))
                        }
                }
        }
        return(shift)
}

get_separator <- function(nucleotide_change){
        if (str_detect(nucleotide_change, "-")) {
                return("-")
        }else if (str_detect(nucleotide_change, "\\*")) {
                return("\\*")
        }else{
                return("\\.")
        }
}

get_indel_separator <- function(nucleotide_change){
        if (str_detect(nucleotide_change, "-")) {
                return("_-")
        }else if (str_detect(nucleotide_change, "\\*")) {
                return("_\\*")
        }else{
                return("_")
        }
}

apply_snp <- function(sequence, SNPs, is_5_prime) {
        
        for (SNP in seq(1, nrow(SNPs))) {
                if(is_5_prime){
                        start <- nchar(sequence)-as.numeric(SNPs[SNP,"pos"])+1
                        end <- nchar(sequence)-as.numeric(SNPs[SNP,"pos"])+1
                }else{
                        start <- as.numeric(SNPs[SNP,"pos"])
                        end <- as.numeric(SNPs[SNP,"pos"])
                }
               str_sub(sequence, start, end) <- SNPs[SNP, "alt"]
        }
        return(sequence)
}

apply_deletion <- function(sequence, DELs, is_5_prime) {
        
        for (DEL in seq(1, nrow(DELs))) {
                if(is_5_prime){
                        start <- nchar(sequence)-as.numeric(DELs[DEL,"pos"])+1
                        end <- nchar(sequence)-as.numeric(DELs[DEL,"pos"])+nchar(DELs[DEL,"ref"])
                }else{
                        start <- as.numeric(DELs[DEL,"pos"])
                        end <- as.numeric(DELs[DEL,"pos"])+nchar(DELs[DEL,"ref"])-1
                }
                str_sub(sequence, start, end) <- paste0(rep("D",nchar(DELs[DEL,"ref"])), collapse = "")
        }
        return(sequence)
        
}

apply_insertion <- function(sequence, INs, is_5_prime) {
        # Apply deletions from last (most downstream) to first (upstream)
        INs <- INs %>% arrange(pos)
        for (IN in seq(1, nrow(INs))) {
                if(is_5_prime){
                        start <- nchar(sequence) - as.numeric(INs[IN,"pos"]) + 1
                }else{
                        start <- as.numeric(INs[IN,"pos"])
                }
                sequence <- paste0(str_sub(sequence, 1, start),
                                   INs[IN, "alt"],
                                   str_sub(sequence, start+1, nchar(sequence)))
        }
        
        return(sequence)
        
}

#### MAIN ####

# load transcript sequences, annotations, samples list and vcf file
sequences <- read_csv(snakemake@input[["wt_cds"]])
protein_coding_transcripts <- read_csv(snakemake@input[["annotations"]])
haplotype_ID <- read_csv(snakemake@input[["haplotypes"]])
haplotype_ID <- haplotype_ID %>% column_to_rownames("haplotype_transcript_id")
# Extract chromosome name from haplotype filename and read threads from snakemake
chr <- snakemake@input[["haplotypes"]] %>% str_split_i(pattern = "chr", 2) %>% str_split_i(pattern = ".csv", 1)
cores <- snakemake@threads

# Subset wild type sequences and transcript annotations only to genes in the current chromosome
protein_coding_transcripts <- protein_coding_transcripts %>% filter(chromosome_name == chr)
sequences <- sequences %>% filter(ensembl_transcript_id %in% unique(protein_coding_transcripts$ensembl_transcript_id))

# Use ensembl transcript id as rowname
sequences <- sequences %>% column_to_rownames(var = "ensembl_transcript_id")
res <- mclapply(rownames(haplotype_ID), FUN = function(haplotype){
                # Only compute sequence if haplotype is not wt
                if(!str_detect(haplotype, "\\.0")){
                        return(cbind(data.frame("haplotype_transcript_id" = haplotype),
                                     build_mutated_sequence(sequences[str_split_i(haplotype, "\\-", 1),],
                                                            haplotype_ID[haplotype, c("variant_types", "nucleotide_change")],
                                                            haplotype
                                                            )
                                     )
                               )
                }else{
                        return(data.frame("haplotype_transcript_id" = haplotype,
                                          "nn_sequence" = sequences[str_split_i(haplotype, "\\-", 1),"coding"],
                                          "start_gain" = FALSE,
                                          "multiple_starts_in_5" = FALSE,
                                          "stop_loss_3_translated" = FALSE)
                               )
                }
        },
        mc.preschedule = TRUE,
        mc.cores = cores,
        mc.cleanup = FALSE
)
res <- bind_rows(res)

haplotype_ID <- haplotype_ID %>% rownames_to_column(var = "haplotype_transcript_id") %>% left_join(res) %>% relocate(nn_sequence, .after = stop_loss_3_translated)

write_csv(haplotype_ID, file = snakemake@output[[1]])
